lib/ex_aws/auth.ex

```diff
@@ -30,10 +30,19 @@ defmodule ExAws.Auth do
   def presigned_url(http_method, url, service, datetime, config, expires, query_params \\ []) do
     service = service_name(service)
     headers = presigned_url_headers(url)
-    query = presigned_url_query(service, datetime, config, expires, query_params)
-    url = "#{url}?#{query}"
-    signature = signature(http_method, url, headers, nil, service, datetime, config)
-    "#{url}&X-Amz-Signature=#{signature}"
+
+    org_query_params = query_params |> Enum.map(fn({k, v}) -> {to_string(k), v} end)
+    amz_query_params = build_amz_query_params(service, datetime, config, expires)
+    {org_query, amz_query} = [org_query_params, amz_query_params]
+    |> Enum.map(&(canonical_query_params(&1)))
+    |> List.to_tuple
+    query_to_sign = org_query_params ++ amz_query_params |> canonical_query_params
+    query_for_url = if Enum.count(org_query_params) == 0 do amz_query else org_query <> "&" <> amz_query end
```

Comment:
Thank you for your readable and better code. I fixed it. https://github.com/CargoSense/ex_aws/pull/233/commits/bba05c30569408ad9b4950a829b8861408998e71